plugin/pkg/admission/runtimeclass/admission.go

```diff
@@ -0,0 +1,199 @@
+	return &RuntimeClass{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+}
+
+// admissionAction handles Admit and Validate phases of admission, switching based on the admissionPhase parameter
+func admissionAction(r *RuntimeClass, attributes admission.Attributes, admissionPhase string) (err error) {
+
+	// Ignore all calls to subresources or resources other than pods.
+	if shouldIgnore(attributes) {
+		return nil
+	}
+	pod, ok := attributes.GetObject().(*api.Pod)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+
+	// Get RuntimeClass. Not finding a given runtimeClass does not return an error. We will want to continue in the
+	// case of Validate phase, even if a RuntimeClass is not defined
+	runtimeClass, err := getRuntimeClass(r, pod, pod.Spec.RuntimeClassName)
+	if err != nil {
+		return err
+	}
+
+	switch admissionPhase {
+	case "admit":
+		err := setOverhead(pod, runtimeClass)
+		if err != nil {
+			return err
+		}
+
+	case "validate":
+		err := validateOverhead(pod, runtimeClass)
+		if err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("invalid value for admissionPhase: %s", admissionPhase)
+	}
+
+	return nil
+}
+
+// getRuntimeClass will return a reference to the RuntimeClass object if it is found. If it cannot be found, or a RuntimeClassName
+// is not provided in the pod spec, *node.RuntimeClass returned will be nil
+func getRuntimeClass(r *RuntimeClass, pod *api.Pod, runtimeClassName *string) (*node.RuntimeClass, error) {
+
+	runtimeClass := &node.RuntimeClass{}
+
+	if runtimeClassName != nil {
+		v1beta1runtimeClass, _ := r.lister.Get(*runtimeClassName)
+		if v1beta1runtimeClass != nil {
+			err := nodev1beta1.Convert_v1beta1_RuntimeClass_To_node_RuntimeClass(v1beta1runtimeClass, runtimeClass, nil)
+			if err != nil {
+				return runtimeClass, err
+			}
+		}
+	}
+	return runtimeClass, nil
+}
+
+func setOverhead(pod *api.Pod, runtimeClass *node.RuntimeClass) (err error) {
+
+	if pod.Spec.Overhead != nil {
+		return fmt.Errorf("pod rejected: Pod Overhead already set")
+	}
+
+	if runtimeClass != nil {
+		if runtimeClass.Overhead != nil {
+			pod.Spec.Overhead = runtimeClass.Overhead.PodFixed
+		}
+	}
+
+	return nil
+}
+
+func validateOverhead(pod *api.Pod, runtimeClass *node.RuntimeClass) (err error) {
+
+	if runtimeClass != nil && runtimeClass.Overhead != nil {
+		// If the Overhead set doesn't match what is provided in the RuntimeClass definition, reject the pod
+		if !apiequality.Semantic.DeepEqual(runtimeClass.Overhead.PodFixed, pod.Spec.Overhead) {
+			return fmt.Errorf("pod rejected: Pod's Overhead doesn't match RuntimeClass's defined Overhead")
+		}
+	} else {
+		// If RuntimeClass with Overhead is not defined but an Overhead is set for pod, reject the pod
+		if pod.Spec.Overhead != nil {
+			return fmt.Errorf("pod rejected: Pod Overhead set without corresponding RuntimeClass defined Overhead")
```

Comment:
return forbidden api-status error here or the apiserver will interpret the string-error as `500` http status

Reply:
ACK

Reply:
update - PTAL